Shared/src/com/pacoapp/paco/shared/util/QueryJsonParser.java

```diff
@@ -47,68 +48,77 @@ public static SQLQuery parseSqlQueryFromJson(String queryJson, boolean enableGrp
-        }
-
-        if (queryCriteria.has(Constants.VALUES)) {
-          JSONArray cv = queryCriteria.getJSONArray(Constants.VALUES);
-          criteriaValues = new String[cv.length()];
-          for (int i = 0; i < cv.length(); i++) {
-            // identify Json string which could be marked with single or double quotes
-            // to ones with single quotes, because jsql parser considers only values within single quotes
-            // as string value
-            if(cv.get(i).getClass().getName().equals("java.lang.String")) {
-              criteriaValues[i] = Constants.SINGLE_QUOTE+cv.getString(i)+Constants.SINGLE_QUOTE;
-            } else {
-              criteriaValues[i] = cv.getString(i);
+      } 
+              
+      if (crKey.equalsIgnoreCase(Constants.QUERY)) {
+        JSONObject queryCriteria = queryObj.getJSONObject(crKey);
+        if (queryCriteria != null) {
+          Iterator<String> qcItr = queryCriteria.keys();
+          while (qcItr.hasNext()) {
+            String crQcKey = qcItr.next();
+            if (crQcKey.equalsIgnoreCase(Constants.CRITERIA)) {
+              sqlBldr.criteriaQuery(queryCriteria.getString(crQcKey).trim());
+            }
+    
+            if (crQcKey.equalsIgnoreCase(Constants.VALUES)) {
+              JSONArray cv = queryCriteria.getJSONArray(crQcKey);
+              criteriaValues = new String[cv.length()];
+              for (int i = 0; i < cv.length(); i++) {
+                // identify Json string which could be marked with single or double quotes
+                // to ones with single quotes, because jsql parser considers only values within single quotes
+                // as string value
+                if(cv.get(i).getClass().getName().equals("java.lang.String")) {
+                  criteriaValues[i] = Constants.SINGLE_QUOTE+cv.getString(i)+Constants.SINGLE_QUOTE;
+                } else {
+                  criteriaValues[i] = cv.getString(i);
+                }
+              }
+              sqlBldr.criteriaValues(criteriaValues);
             }
           }
-          sqlBldr.criteriaValues(criteriaValues);
         }
       }
-    }
-    
-    if (queryObj.has(Constants.ORDER)) {
-      sqlBldr.sortBy(queryObj.getString(Constants.ORDER).trim());
-    }
-    
-    if (queryObj.has(Constants.LIMIT)) {
-      sqlBldr.limit(queryObj.getString(Constants.LIMIT).trim());
-    }
-    
-    // groupBy feature should be enabled and only if we have group clause, should we have the having column
-    if (enableGrpByAndProjection && queryObj.has(Constants.GROUP)) {
-      sqlBldr.groupBy(queryObj.getString(Constants.GROUP).trim());
-      // This gets set twice. Once when there is a *, and next when there is a group by.
-      // Group by takes higher precedence.
-      sqlBldr.fullEventAndOutputs(false);
-
-      if (queryObj.has(Constants.HAVING)) {
-        sqlBldr.having(queryObj.getString(Constants.HAVING).trim());
+      
+      if (crKey.equalsIgnoreCase(Constants.ORDER)) {
```

Comment:
Micro optimization question: Would it be better to do crKey= crKey.toLowercase() at the beginning of all of these checks? Not sure if equals ignore case has to rebuild the lowercase version each time or if it has a performance impact at all.

Reply:
Yep. Definitely better. Fixed it now.